SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -0,0 +1,191 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSReadReceiptManager.h"
+#import "OWSMessageSender.h"
+#import "OWSReadReceipt.h"
+#import "OWSReadReceiptsMessage.h"
+#import "TSContactThread.h"
+#import "TSDatabaseView.h"
+#import "TSIncomingMessage.h"
+#import "TextSecureKitEnv.h"
+#import "Threading.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface OWSReadReceiptManager ()
+
+@property (nonatomic, readonly) TSStorageManager *storageManager;
+@property (nonatomic, readonly) OWSMessageSender *messageSender;
+
+// A map of "thread unique id"-to-"read receipt" for read receipts that
+// we will send to our linked devices.
+//
+// Should only be accessed while synchronized on the OWSReadReceiptManager.
+@property (nonatomic, readonly) NSMutableDictionary<NSString *, OWSReadReceipt *> *toLinkedDevicesReadReceiptMap;
+
+// Should only be accessed while synchronized on the OWSReadReceiptManager.
+@property (nonatomic) BOOL isProcessing;
+
+@end
+
+#pragma mark -
+
+@implementation OWSReadReceiptManager
+
++ (instancetype)sharedManager
+{
+    static OWSReadReceiptManager *sharedMyManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedMyManager = [[self alloc] initDefault];
+    });
+    return sharedMyManager;
+}
+
+- (instancetype)initDefault
+{
+    OWSMessageSender *messageSender = [TextSecureKitEnv sharedEnv].messageSender;
+
+    return [self initWithMessageSender:messageSender];
+}
+
+- (instancetype)initWithMessageSender:(OWSMessageSender *)messageSender
+{
+    self = [super init];
+
+    if (!self) {
+        return self;
+    }
+
+    _messageSender = messageSender;
+
+    _toLinkedDevicesReadReceiptMap = [NSMutableDictionary new];
+
+    OWSSingletonAssert();
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(databaseViewRegistrationComplete)
+                                                 name:kNSNotificationName_DatabaseViewRegistrationComplete
+                                               object:nil];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)databaseViewRegistrationComplete
+{
+    [self scheduleProcessing];
```

Comment:
Do we need to manually kick this off if db view registration completes before this class is `init`?

Eventually we'll read a message locally, which also kicks it off, but is there a case where we'll still want this to run?

Reply:
Great point!